reconcile/utils/runtime/integration.py

```diff
@@ -0,0 +1,140 @@
+from abc import (
+    ABC,
+    abstractmethod,
+)
+from dataclasses import dataclass
+from types import ModuleType
+from typing import (
+    Any,
+    Callable,
+    Optional,
+)
+
+
+@dataclass
+class ShardedRunProposal:
+
+    proposed_shards: set[str]
+
+
+@dataclass
+class DesiredStateShardConfig:
+    # a callback function that allows the integration to review the proposed
+    # shards and decide if they should be processed with sharded runs or not
+    sharded_run_review: Callable[[ShardedRunProposal], bool]
+    # a set of jsonpath selectors defining where in the desired state offered
+    # by `QontractReconcileIntegration.get_early_exit_desired_state` the shards
+    # can be found
+    shard_path_selectors: set[str]
+    # the name of the integrations `QontractReconcileIntegrationrun` function
+    # argument that is used to pass the shard to process
+    shard_arg_name: str
+    # some integrations offer the shard arg as a colletion of values
+    shard_arg_is_collection: bool = False
+
+
+class QontractReconcileIntegration(ABC):
+    """
+    The base class for all integrations. It defines the basic interface to interact
```

Comment:
+1 to making this a class and `ABC`. Definitely time to start using this pattern moving forward.